DATA RACE in TestObtainCredentialsFromMemberCluster

#### Which jobs are flaking:

https://github.com/karmada-io/karmada/actions/runs/26215493365/job/77137951051

#### Which test(s) are flaking:

TestObtainCredentialsFromMemberCluster

pkg/util/credential_test.go:151

#### Reason for failure:

The failure is caused by a data race on *testing.T due to a leaked background goroutine from `TestObtainCredentialsFromMemberCluster`.

  In `pkg/util/credential_test.go`, the test starts a goroutine that runs `wait.PollUntilContextCancel(...)` and calls `t.Log/t.Logf` inside the poll function. When the subtest finishes, the test framework starts tearing down the *testing.T (its internal log destination/state changes). But
  the goroutine may still be running and still calling t.Log, so:

   - one goroutine reads testing.T internals during t.Log()
   - another goroutine (the test runner) writes/updates them during test cleanup

  Race detector reports this, and it may appear under a different test name because the race is detected while another test is running.

#### Anything else we need to know:

```log
=== Failed
=== FAIL: pkg/util TestRegisterClusterInControllerPlane/report_secret_not_enabled (0.03s)
==================
WARNING: DATA RACE
Read at 0x00c0014fee4b by goroutine 203:
  testing.(*common).destination()
      /opt/hostedtoolcache/go/1.25.7/x64/src/testing/testing.go:1055 +0x1cc
  testing.(*common).log()
      /opt/hostedtoolcache/go/1.25.7/x64/src/testing/testing.go:1027 +0xbe
  testing.(*common).Log()
      /opt/hostedtoolcache/go/1.25.7/x64/src/testing/testing.go:1180 +0x76
  github.com/karmada-io/karmada/pkg/util.TestObtainCredentialsFromMemberCluster.func2.1.1()
      /home/runner/work/karmada/karmada/pkg/util/credential_test.go:148 +0x12d5
  k8s.io/apimachinery/pkg/util/wait.loopConditionUntilContext.func2()
      /home/runner/work/karmada/karmada/vendor/k8s.io/apimachinery/pkg/util/wait/loop.go:87 +0xc5
  k8s.io/apimachinery/pkg/util/wait.loopConditionUntilContext()
      /home/runner/work/karmada/karmada/vendor/k8s.io/apimachinery/pkg/util/wait/loop.go:88 +0x32f
  k8s.io/apimachinery/pkg/util/wait.PollUntilContextCancel()
      /home/runner/work/karmada/karmada/vendor/k8s.io/apimachinery/pkg/util/wait/poll.go:33 +0x6a
  github.com/karmada-io/karmada/pkg/util.TestObtainCredentialsFromMemberCluster.func2.1()
      /home/runner/work/karmada/karmada/pkg/util/credential_test.go:107 +0xb6
  github.com/karmada-io/karmada/pkg/util.TestObtainCredentialsFromMemberCluster.func2.gowrap1()
      /home/runner/work/karmada/karmada/pkg/util/credential_test.go:151 +0x4f

Previous write at 0x00c0014fee4b by goroutine 198:
  testing.tRunner.func1()
      /opt/hostedtoolcache/go/1.25.7/x64/src/testing/testing.go:1921 +0x904
  runtime.deferreturn()
      /opt/hostedtoolcache/go/1.25.7/x64/src/runtime/panic.go:589 +0x5d
  testing.(*T).Run.gowrap1()
      /opt/hostedtoolcache/go/1.25.7/x64/src/testing/testing.go:1997 +0x44
make: *** [Makefile:123: test] Error 1

Goroutine 203 (running) created at:
  github.com/karmada-io/karmada/pkg/util.TestObtainCredentialsFromMemberCluster.func2()
      /home/runner/work/karmada/karmada/pkg/util/credential_test.go:106 +0x1a4
  github.com/karmada-io/karmada/pkg/util.TestObtainCredentialsFromMemberCluster.func3()
      /home/runner/work/karmada/karmada/pkg/util/credential_test.go:169 +0x143
  testing.tRunner()
      /opt/hostedtoolcache/go/1.25.7/x64/src/testing/testing.go:1934 +0x21c
  testing.(*T).Run.gowrap1()
      /opt/hostedtoolcache/go/1.25.7/x64/src/testing/testing.go:1997 +0x44

Goroutine 198 (finished) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.25.7/x64/src/testing/testing.go:1997 +0x9d2
  testing.runTests.func1()
      /opt/hostedtoolcache/go/1.25.7/x64/src/testing/testing.go:2477 +0x85
  testing.tRunner()
      /opt/hostedtoolcache/go/1.25.7/x64/src/testing/testing.go:1934 +0x21c
  testing.runTests()
      /opt/hostedtoolcache/go/1.25.7/x64/src/testing/testing.go:2475 +0x96c
  testing.(*M).Run()
      /opt/hostedtoolcache/go/1.25.7/x64/src/testing/testing.go:2337 +0xed4
  main.main()
      _testmain.go:269 +0x164
==================
    testing.go:1617: race detected during execution of test
    --- FAIL: TestRegisterClusterInControllerPlane/report_secret_not_enabled (0.03s)

=== FAIL: pkg/util TestRegisterClusterInControllerPlane (0.08s)
```

--- a/pkg/util/credential_test.go
+++ b/pkg/util/credential_test.go
@@ -19,6 +19,7 @@ package util
 import (
 	"context"
 	"reflect"
+	"sync"
 	"testing"
 	"time"
 
@@ -59,7 +60,10 @@ func TestObtainCredentialsFromMemberCluster(t *testing.T) {
 				},
 				aop: func(t *testing.T, clusterKubeClient kubernetes.Interface) func() {
 					ctx, cancel := context.WithCancel(context.TODO())
+					var wg sync.WaitGroup
+					wg.Add(1)
 					go func(clusterKubeClient kubernetes.Interface) {
+						defer wg.Done()
 						_ = wait.PollUntilContextCancel(ctx, 1*time.Second, true, func(ctx context.Context) (done bool, err error) {
 							impersonationSA, err := clusterKubeClient.CoreV1().ServiceAccounts("karmada-cluster").Get(ctx, names.GenerateServiceAccountName("impersonator"), metav1.GetOptions{})
 							if err != nil {
@@ -86,6 +90,7 @@ func TestObtainCredentialsFromMemberCluster(t *testing.T) {
 					}(clusterKubeClient)
 					return func() {
 						cancel()
+						wg.Wait()
 					}
 				},
 			},
@@ -103,7 +108,10 @@ func TestObtainCredentialsFromMemberCluster(t *testing.T) {
 				},
 				aop: func(t *testing.T, clusterKubeClient kubernetes.Interface) func() {
 					ctx, cancel := context.WithCancel(context.TODO())
+					var wg sync.WaitGroup
+					wg.Add(1)
 					go func(clusterKubeClient kubernetes.Interface) {
+						defer wg.Done()
 						_ = wait.PollUntilContextCancel(ctx, 1*time.Second, true, func(ctx context.Context) (done bool, err error) {
 							impersonationSA, err := clusterKubeClient.CoreV1().ServiceAccounts("karmada-cluster").Get(ctx, names.GenerateServiceAccountName("impersonator"), metav1.GetOptions{})
 							if err != nil {
@@ -151,6 +159,7 @@ func TestObtainCredentialsFromMemberCluster(t *testing.T) {
 					}(clusterKubeClient)
 					return func() {
 						cancel()
+						wg.Wait()
 					}
 				},
 			},

